File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    result = []
    
    todaylist = today.split(".")
    [todayyear, todaymonth, todaydate] = [int(todaylist[0]), int(todaylist[1]), int(todaylist[2])]
    termsdict = {i.split()[0]: int(i.split()[1]) for i in terms}
    
    for num in range(len(privacies)):
        privacydate = privacies[num].split()[0].split(".")
        privacyterm = termsdict[privacies[num].split()[1]]
        privacydate[0] = int(privacydate[0])
        privacydate[1] = int(privacydate[1])+privacyterm 
        privacydate[2] = int(privacydate[2]) - 1
        if privacydate[2] < 1:
            privacydate[1] -=1
            privacydate[2] = 28

        if privacydate[1] > 12:
            if privacydate[1]%12 == 0:
                privacydate[0] = privacydate[0] + int(privacydate[1] / 12) - 1
                privacydate[1] = 12
            else:
                privacydate[0] += int(privacydate[1] / 12)
                privacydate[1] = privacydate[1] % 12

        if (privacydate[0] < todayyear) or ((privacydate[0] == todayyear) and (privacydate[1] < todaymonth)) or ((privacydate[0] == todayyear) and (privacydate[1] == todaymonth) and (privacydate[2] < todaydate)):
            result.append(num+1)
    
    return result

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;
class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        
        String[] today1 = today.split("\\.");
        int todayYear = Integer.parseInt(today1[0]);
        int todayMonth = Integer.parseInt(today1[1]);
        int todayDate = Integer.parseInt(today1[2]);
        
        Map<String, Integer> termsMap = new HashMap<>();
        
        for (int i = 0; i < terms.length; i++) {
            String[] personalTerm = terms[i].split(" ");
            termsMap.put(personalTerm[0], Integer.parseInt(personalTerm[1]));
        }
        
        List<Integer> result = new ArrayList();
        
        for (int i=0; i<privacies.length; i++) {
            String[] dateAndChar = privacies[i].split(" ");
            String person = dateAndChar[1];
            String[] date = dateAndChar[0].split("\\.");
            
            int fromYear = Integer.parseInt(date[0]);
            int fromMonth = Integer.parseInt(date[1]);
            int fromDate = Integer.parseInt(date[2]);
            
            int lastMonth = fromMonth + termsMap.get(person);
            int lastYear = fromYear;
            int lastDate = fromDate;
            
            if (lastMonth > 12 && lastMonth % 12 != 0) {
                lastYear = lastYear + lastMonth/12;
                lastMonth = lastMonth % 12;
            } else if (lastMonth % 12 == 0) {
                lastYear = lastYear + lastMonth/12 - 1;
                lastMonth = 12;
            }
            
            if (todayYear > lastYear || (todayYear == lastYear && todayMonth > lastMonth) || (todayYear == lastYear && todayMonth == lastMonth && todayDate >= lastDate)) {
                result.add(i+1);
            }
            
        }
        
        return result.stream().mapToInt(Integer::intValue).toArray();
        
    }
}